### ISSUE TITLE
Unable to activate NDN-DPDK as fileserver

### ISSUE THREAD

**Issue by cataliniordache** [usnistgov/ndn-dpdk#33]
I am trying to test the fileserver, but with no success so far. I am using the latest revision of this repo, and here is the message I get when I try to activate it:

```bash
{"level":"info","ts":1632764246.8898258,"logger":"main","msg":"NDN-DPDK service starting","version":"v0.0.0-20210921182539-d120d8a44426","uid":0,"linux":"5.13.12-1.el8.elrepo.x86_64","dpdk":"DPDK 21.08.0","spdk":"SPDK v21.07"}
{"level":"debug","ts":1632764246.8902588,"logger":"main","msg":"systemd watchdog not configured"}
{"level":"info","ts":1632764246.891023,"logger":"main","msg":"GraphQL HTTP server starting","listen":":3030"}
{"level":"info","ts":1632764553.51484,"logger":"main","msg":"activate start","role":"fileserver"}
{"level":"debug","ts":1632764553.5390573,"logger":"DPDK","msg":"EAL: Detected 32 lcore(s)"}
{"level":"debug","ts":1632764553.5391133,"logger":"DPDK","msg":"EAL: Detected 2 NUMA nodes"}
{"level":"debug","ts":1632764553.539708,"logger":"DPDK","msg":"EAL: Detected shared linkage of DPDK"}
{"level":"debug","ts":1632764553.6047363,"logger":"DPDK","msg":"EAL: Selected IOVA mode 'VA'"}
{"level":"warn","ts":1632764553.604979,"logger":"DPDK","msg":"EAL: No available 2048 kB hugepages reported"}
{"level":"warn","ts":1632764553.6050215,"logger":"DPDK","msg":"EAL: No free 2048 kB hugepages reported on node 0"}
{"level":"warn","ts":1632764553.6050532,"logger":"DPDK","msg":"EAL: No free 2048 kB hugepages reported on node 1"}
{"level":"warn","ts":1632764553.605085,"logger":"DPDK","msg":"EAL: No available 2048 kB hugepages reported"}
{"level":"warn","ts":1632764553.9590797,"logger":"DPDK.8","msg":"TELEMETRY: No legacy callbacks, legacy socket not created"}
{"level":"info","ts":1632764553.9591076,"logger":"ealinit","msg":"EAL ready","args":"-l 8,9,10,11 --socket-limit 1,6442450944 --file-prefix fileserver --in-memory --single-file-segments -d /usr/local/lib/dpdk/pmds-21.3 --no-pci","main":8,"workers":[9,10,11],"sockets":[1]}
{"level":"info","ts":1632764553.9752936,"logger":"SPDK.8","msg":"SPDK logger ready"}
{"level":"debug","ts":1632764553.9940004,"logger":"ealinit","msg":"MainThread is running"}
{"level":"info","ts":1632764553.9976888,"logger":"pktmbuf","msg":"decreasing Dataroom","key":"PAYLOAD","oldDataroom":11285,"newDataroom":9128}
{"level":"debug","ts":1632764553.997762,"logger":"eal","msg":"object ID allocated","type":"ethvdev.Memif","id":"K0000000000000001"}
{"level":"debug","ts":1632764553.9978864,"logger":"DPDK","msg":"rte_pmd_memif_probe(): Initialize MEMIF: net_memifK0000000000000001."}
{"level":"warn","ts":1632764553.997945,"logger":"DPDK","msg":"rte_pmd_memif_probe(): Failed to register mp action callback: Operation not supported"}
{"level":"error","ts":1632764553.9980826,"logger":"DPDK","msg":"RING: Cannot reserve memory"}
{"level":"error","ts":1632764553.998119,"logger":"DPDK","msg":"HASH: memory allocation failed"}
{"level":"error","ts":1632764553.9981751,"logger":"DPDK","msg":"memif_socket_init(): Failed to create memif socket hash."}
{"level":"error","ts":1632764553.9981961,"logger":"eal","msg":"rte_vdev_init error","name":"net_memifK0000000000000001","args":"rsize=10,socket=/run/ndn/fileserver.sock,socket-abstract=no,mac=F2:6D:65:6D:69:66,id=0,role=client,bsize=9000","socket":"any","error":"1 Operation not permitted"}
```

Here is the configuration file for the forwarder (which successfully starts)

```ts
// fw-args.ts
import type { ActivateFwArgs } from "@usnistgov/ndn-dpdk";
import stdout from "stdout-stream";

const args: ActivateFwArgs = {
  eal: {
    cores: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16],
    pciDevices: [
      "0000:5e:00.0",
    ],
  },
  mempool: {
    DIRECT: { capacity: 1048575, dataroom: 9128 },
    INDIRECT: { capacity: 1048575 },
  },
  fib: {
    capacity: 4095,
    startDepth: 8,
  },
  pcct: {
    pcctCapacity: 65535,
    csDirectCapacity: 20000,
    csIndirectCapacity: 20000,
  },
};

stdout.write(JSON.stringify(args));
```

And the configuration file for the fileserver:
```ts
// fileserver-args.ts
import type { ActivateFileServerArgs } from "@usnistgov/ndn-dpdk";
import stdout from "stdout-stream";

const args: ActivateFileServerArgs = {
  eal: {
    coresPerNuma: { 0: 0, 1: 4 },
    memPerNuma: { 0: 0, 1: 2048 },
    filePrefix: "fileserver",
  },
  mempool: {
    DIRECT: { capacity: 65535, dataroom: 9128 },
    INDIRECT: { capacity: 65535 },
    PAYLOAD: { capacity: 65535, dataroom: 9128 },
  },
  face: {
    scheme: "memif",
    socketName: "/run/ndn/fileserver.sock",
    id: 0,
    role: "client",
    dataroom: 9000,
  },
  fileServer: {
    nThreads: 1,
    mounts: [
      { prefix: "/ndnc/ft", path: "/root/samples/" }
    ],
    segmentLen: 6 * 1024,
    uringCapacity: 4096,
  },
};

stdout.write(JSON.stringify(args));

```

Prior to activating the fileserver, I create a new face and a FIB entry for it like this:
```bash
FACEID=$(gq http://172.17.0.2:3030 \
    -q 'mutation createFace($locator:JSON!){createFace(locator:$locator){id}}' \
    --variablesJSON '{
      "locator": {
        "scheme": "memif",
        "socketName": "/run/ndn/fileserver.sock",
        "id": 0,
        "role": "server",
        "dataroom": 9000
      }
    }' | jq -c .data.createFace | tee /dev/stderr | jq -r .id)
```
```bash
docker run --network host -i --rm ndndpdk ndndpdk-ctrl --gqlserver $GQLSERVER insert-fib \
    --name /fileserver --nexthop $FACEID
```

**Comment by yoursunny** [comment]
These are the real errors:

> {"level":"error","ts":1632764553.9980826,"logger":"DPDK","msg":"RING: Cannot reserve memory"}
> {"level":"error","ts":1632764553.998119,"logger":"DPDK","msg":"HASH: memory allocation failed"}

The `--socket-limit` flag also suspicious: you specified 2048 but it became 6442450944, but I can't reproduce this:

> {"level":"info","ts":1632764553.9591076,"logger":"ealinit","msg":"EAL ready","args":"-l 8,9,10,11 --socket-limit 1,6442450944 --file-prefix fileserver --in-memory --single-file-segments -d /usr/local/lib/dpdk/pmds-21.3 --no-pci","main":8,"workers":[9,10,11],"sockets":[1]}

I'm seeing the same "cannot reserve memory" error when I set **.eal.memPerNuma[0]** to zero or a small value.
It seems that DPDK is attempting to create some data structure in NUMA socket 0, even if every CPU is on NUMA socket 1.
Then, the `--socket-limit` setting for that NUMA socket is 1MB, so that data structure creation fails.
I'll find out what's going on.

As a temporary workaround: run the program on NUMA socket 0, or increase **.eal.memPerNuma[0]** to a larger value (try 2048 and adjust from there).

**Comment by yoursunny** [comment]
> It seems that DPDK is attempting to create some data structure in NUMA socket 0, even if every CPU is on NUMA socket 1.

This is likely a DPDK bug.
https://github.com/DPDK/dpdk/blob/a0a388a897167f0efaf1657d9cdc32b14bd83604/drivers/net/memif/memif_socket.c#L939-L950
This function calls `rte_hash_create` without assigning `rte_hash_parameters#socket_id` so it defaults to zero.
DPDK tries to create a hashtable on NUMA socket 0, but it cannot because `--socket-limit` is set to 1MB.
I will either report bug or send patch to DPDK.

**Comment by cataliniordache** [comment]
Thanks, @yoursunny. Setting 2048 instead of 0 for NUMA 0 as well, made things progress a bit, but I am still having some issues while activating the fileserver. The command returns this error message:
```bash
2021/09/28 16:42:21 graphql: runtime error: invalid memory address or nil pointer dereference
```

while the output is this:
```bash
{"level":"debug","ts":1632847173.0871453,"logger":"main","msg":"systemd watchdog not configured"}
{"level":"info","ts":1632847173.0877752,"logger":"main","msg":"GraphQL HTTP server starting","listen":":3030"}
{"level":"info","ts":1632847340.5309494,"logger":"main","msg":"activate start","role":"fileserver"}
{"level":"debug","ts":1632847340.5542536,"logger":"DPDK","msg":"EAL: Detected 32 lcore(s)"}
{"level":"debug","ts":1632847340.55431,"logger":"DPDK","msg":"EAL: Detected 2 NUMA nodes"}
{"level":"debug","ts":1632847340.554853,"logger":"DPDK","msg":"EAL: Detected shared linkage of DPDK"}
{"level":"debug","ts":1632847340.6206298,"logger":"DPDK","msg":"EAL: Selected IOVA mode 'VA'"}
{"level":"info","ts":1632847340.6208608,"logger":"DPDK","msg":"EAL: 4092 hugepages of size 2097152 reserved, but no mounted hugetlbfs found for that size"}
{"level":"info","ts":1632847340.6210458,"logger":"DPDK","msg":"EAL: 32 hugepages of size 1073741824 reserved, but no mounted hugetlbfs found for that size"}
{"level":"warn","ts":1632847340.7385905,"logger":"DPDK.8","msg":"TELEMETRY: No legacy callbacks, legacy socket not created"}
{"level":"info","ts":1632847340.7424254,"logger":"ealinit","msg":"EAL ready","args":"-l 8,9,10,11 --socket-limit 2048,2048 --file-prefix fileserver --in-memory --single-file-segments -d /usr/local/lib/dpdk/pmds-21.3 --no-pci","main":8,"workers":[9,10,11],"sockets":[1]}
{"level":"info","ts":1632847340.7543142,"logger":"SPDK.8","msg":"SPDK logger ready"}
{"level":"debug","ts":1632847341.0194066,"logger":"ealinit","msg":"MainThread is running"}
{"level":"info","ts":1632847341.0279489,"logger":"pktmbuf","msg":"decreasing Dataroom","key":"PAYLOAD","oldDataroom":11285,"newDataroom":9128}
{"level":"debug","ts":1632847341.0280397,"logger":"eal","msg":"object ID allocated","type":"ethvdev.Memif","id":"K0000000000000001"}
{"level":"info","ts":1632847341.0290995,"logger":"eal","msg":"vdev initialized","name":"net_memifK0000000000000001","args":"socket=/run/ndn/fileserver.sock,socket-abstract=no,mac=F2:6D:65:6D:69:66,id=0,role=client,bsize=9000,rsize=10","socket":"any"}
{"level":"info","ts":1632847341.0291607,"logger":"ethface","msg":"port opened","port":0}
{"level":"debug","ts":1632847341.0290904,"logger":"DPDK","msg":"rte_pmd_memif_probe(): Initialize MEMIF: net_memifK0000000000000001."}
{"level":"warn","ts":1632847341.02925,"logger":"DPDK","msg":"rte_pmd_memif_probe(): Failed to register mp action callback: Operation not supported"}
{"level":"debug","ts":1632847341.02973,"logger":"eal","msg":"object ID allocated","type":"pktmbuf.Pool","id":"K0000000000000002"}
{"level":"debug","ts":1632847341.0528362,"logger":"pktmbuf","msg":"mempool created","template":"DIRECT","socket":1,"use-socket":"any","cfg":{"capacity":65535,"privSize":480,"dataroom":9128},"pool":"K0000000000000002"}
{"level":"debug","ts":1632847341.0528786,"logger":"eal","msg":"object ID allocated","type":"pktmbuf.Pool","id":"K0000000000000003"}
{"level":"debug","ts":1632847341.063283,"logger":"pktmbuf","msg":"mempool created","template":"INDIRECT","socket":1,"use-socket":"any","cfg":{"capacity":65535,"privSize":0,"dataroom":0},"pool":"K0000000000000003"}
{"level":"debug","ts":1632847341.0633268,"logger":"eal","msg":"object ID allocated","type":"pktmbuf.Pool","id":"K0000000000000004"}
{"level":"debug","ts":1632847341.0792909,"logger":"pktmbuf","msg":"mempool created","template":"HEADER","socket":1,"use-socket":"any","cfg":{"capacity":65535,"privSize":480,"dataroom":186},"pool":"K0000000000000004"}
{"level":"debug","ts":1632847341.0793116,"logger":"eal","msg":"object ID allocated","type":"ringbuffer.Ring","id":"K0000000000000005"}
{"level":"debug","ts":1632847341.0793371,"logger":"eal","msg":"object ID allocated","type":"iface.Reassembler","id":"K0000000000000006"}
{"level":"debug","ts":1632847341.0793648,"logger":"eal","msg":"object ID allocated","type":"iface.Reassembler","id":"K0000000000000007"}
{"level":"debug","ts":1632847341.0794032,"logger":"eal","msg":"object ID allocated","type":"iface.Reassembler","id":"K0000000000000008"}
{"level":"debug","ts":1632847341.0794196,"logger":"eal","msg":"object ID allocated","type":"iface.Reassembler","id":"K0000000000000009"}
{"level":"debug","ts":1632847341.0794413,"logger":"eal","msg":"object ID allocated","type":"iface.Reassembler","id":"K000000000000000a"}
{"level":"debug","ts":1632847341.0794573,"logger":"eal","msg":"object ID allocated","type":"iface.Reassembler","id":"K000000000000000b"}
{"level":"debug","ts":1632847341.0794787,"logger":"eal","msg":"object ID allocated","type":"iface.Reassembler","id":"K000000000000000c"}
{"level":"debug","ts":1632847341.079494,"logger":"eal","msg":"object ID allocated","type":"iface.Reassembler","id":"K000000000000000d"}
{"level":"info","ts":1632847341.079531,"logger":"ethface","msg":"impl initialized","port":0,"impl":"RxMemif"}
{"level":"debug","ts":1632847341.0795565,"logger":"pktmbuf","msg":"mempool found","template":"DIRECT","socket":"any","use-socket":"any","cfg":{"capacity":65535,"privSize":480,"dataroom":9128},"pool":"K0000000000000002"}
{"level":"info","ts":1632847341.0796874,"logger":"ethdev","msg":"ethdev started","id":0,"name":"net_memifK0000000000000001","mtu":"unchanged","rxq":1,"txq":1,"promisc":false}
{"level":"info","ts":1632847341.0797238,"logger":"ethface","msg":"face started","port":0,"impl":"RxMemif","face":44995}
{"level":"info","ts":1632847341.0797465,"logger":"iface","msg":"face created","id":44995,"socket":1,"mtu":9014,"locator":{"rxQueueSize":64,"txQueueSize":64,"role":"client","socketName":"/run/ndn/fileserver.sock","id":0,"dataroom":9000,"ringCapacity":1024}}
{"level":"debug","ts":1632847341.0798645,"logger":"eal","msg":"object ID allocated","type":"ringbuffer.Ring","id":"K000000000000000e"}
{"level":"debug","ts":1632847341.0799394,"logger":"eal","msg":"object ID allocated","type":"pktmbuf.Pool","id":"K000000000000000f"}
{"level":"debug","ts":1632847341.3038888,"logger":"DPDK","msg":"memif_connect(): Connected."}
{"level":"debug","ts":1632847341.303959,"logger":"DPDK","msg":"memif_msg_receive_connected(): Remote interface net_memifK0000000000000049 connected."}
{"level":"debug","ts":1632847341.322807,"logger":"pktmbuf","msg":"mempool created","template":"PAYLOAD","socket":1,"use-socket":"any","cfg":{"capacity":65535,"privSize":480,"dataroom":9128},"pool":"K000000000000000f"}
{"level":"debug","ts":1632847341.3228931,"logger":"eal","msg":"object ID allocated","type":"ringbuffer.Ring","id":"K0000000000000010"}
{"level":"debug","ts":1632847341.322942,"logger":"pktmbuf","msg":"mempool found","template":"PAYLOAD","socket":1,"use-socket":"any","cfg":{"capacity":65535,"privSize":480,"dataroom":9128},"pool":"K000000000000000f"}
```

Now the config file looks like this:
```ts
// fileserver-args.ts
import type { ActivateFileServerArgs } from "@usnistgov/ndn-dpdk";
import stdout from "stdout-stream";

const args: ActivateFileServerArgs = {
  eal: {
    coresPerNuma: { 0: 0, 1: 4 },
    memPerNuma: { 0: 2048, 1: 2048 },
    filePrefix: "fileserver",
  },
  mempool: {
    DIRECT: { capacity: 65535, dataroom: 9128 },
    INDIRECT: { capacity: 65535 },
    PAYLOAD: { capacity: 65535, dataroom: 9128 },
  },
  face: {
    scheme: "memif",
    socketName: "/run/ndn/fileserver.sock",
    id: 0,
    role: "client",
    dataroom: 9000,
  },
  fileServer: {
    nThreads: 1,
    mounts: [
      { prefix: "/ndnc/ft", path: "/root/samples" }
    ],
    segmentLen: 6 * 1024,
    uringCapacity: 4096,
  },
};

stdout.write(JSON.stringify(args));
```

Any thoughts?

**Comment by Pesa** [comment]
I don't see any errors in the logs above. What makes you say that there are still issues with the activation?

**Comment by cataliniordache** [comment]
The command output: `graphql: runtime error: invalid memory address or nil pointer dereference` and the fact that the routing isn't working, meaning that I am not sure if Interests are reaching the server, but for sure I don't get any Data back to the consumer

**Comment by Pesa** [comment]
oh sorry, I completely missed that line in your message. @yoursunny any ideas?

**Comment by yoursunny** [comment]
> > It seems that DPDK is attempting to create some data structure in NUMA socket 0, even if every CPU is on NUMA socket 1.
> 
> This is likely a DPDK bug. I will either report bug or send patch to DPDK.

DPDK patch sent, and it's included in `ndndpdk-depends.sh`.
https://patches.dpdk.org/project/dpdk/patch/973fc5b37c6164c1@cs.arizona.edu/

> runtime error: invalid memory address or nil pointer dereference

This might be a bug in Go code: something called by the GraphQL resolver is using a nil pointer.
However, I don't know where's the problem.

> {"level":"info","ts":1632847340.6208608,"logger":"DPDK","msg":"EAL: 4092 hugepages of size 2097152 reserved, but no mounted hugetlbfs found for that size"}
> {"level":"info","ts":1632847340.6210458,"logger":"DPDK","msg":"EAL: 32 hugepages of size 1073741824 reserved, but no mounted hugetlbfs found for that size"}

This is somewhat concerning. hugepages are not mounted?

>     memPerNuma: { 0: 2048, 1: 2048 },

With the DPDK patch mentioned above, it's unnecessary to allocate memory on both NUMA sockets.
I'm sometimes seeing allocation errors when limited to 2048MB.
Now the recommendation in `docs/activate/fileserver-args.ts` is 4096MB.
You can try change this to `memPerNuma: { 0: 0, 1: 4096 },`.

**Comment by Pesa** [comment]
> > runtime error: invalid memory address or nil pointer dereference
> 
> This might be a bug in Go code: something called by the GraphQL resolver is using a nil pointer. However, I don't know where's the problem.

How can we obtain more info such as a stack trace?

**Comment by yoursunny** [comment]
> > > runtime error: invalid memory address or nil pointer dereference
> > 
> > 
> > This might be a bug in Go code: something called by the GraphQL resolver is using a nil pointer. However, I don't know where's the problem.
> 
> How can we obtain more info such as a stack trace?

This is blocked by https://github.com/graphql-go/graphql/issues/243

**Comment by cataliniordache** [comment]
I did create a new Docker image, but still, I don't get the Data back from the server, and I don't know if the Interest reaches the server. Are there any logs for this that I can activate?

Here is the current output of the fileserver app:
```bash
{"level":"info","ts":1632937740.3103666,"logger":"main","msg":"NDN-DPDK service starting","version":"v0.0.0-20210929141028-d4b66d653742","uid":0,"linux":"5.13.12-1.el8.elrepo.x86_64","dpdk":"DPDK 21.08.0","spdk":"SPDK v21.07"}
{"level":"debug","ts":1632937740.3109243,"logger":"main","msg":"systemd watchdog not configured"}
{"level":"info","ts":1632937740.3115845,"logger":"main","msg":"GraphQL HTTP server starting","listen":":3030"}
{"level":"info","ts":1632937783.3358536,"logger":"main","msg":"activate start","role":"fileserver"}
{"level":"debug","ts":1632937783.3593972,"logger":"DPDK","msg":"EAL: Detected 32 lcore(s)"}
{"level":"debug","ts":1632937783.3594506,"logger":"DPDK","msg":"EAL: Detected 2 NUMA nodes"}
{"level":"debug","ts":1632937783.360031,"logger":"DPDK","msg":"EAL: Detected shared linkage of DPDK"}
{"level":"debug","ts":1632937783.432585,"logger":"DPDK","msg":"EAL: Selected IOVA mode 'VA'"}
{"level":"warn","ts":1632937783.4328213,"logger":"DPDK","msg":"EAL: No available 2048 kB hugepages reported"}
{"level":"warn","ts":1632937783.432855,"logger":"DPDK","msg":"EAL: No free 2048 kB hugepages reported on node 0"}
{"level":"warn","ts":1632937783.4328923,"logger":"DPDK","msg":"EAL: No free 2048 kB hugepages reported on node 1"}
{"level":"warn","ts":1632937783.43292,"logger":"DPDK","msg":"EAL: No available 2048 kB hugepages reported"}
{"level":"warn","ts":1632937783.7865424,"logger":"DPDK.8","msg":"TELEMETRY: No legacy callbacks, legacy socket not created"}
{"level":"info","ts":1632937783.7865674,"logger":"ealinit","msg":"EAL ready","args":"-l 8,9,10,11 --socket-limit 1,4096 --file-prefix fileserver --in-memory --single-file-segments -d /usr/local/lib/dpdk/pmds-21.3 --no-pci","main":8,"workers":[9,10,11],"sockets":[1]}
{"level":"info","ts":1632937783.8028982,"logger":"SPDK.8","msg":"SPDK logger ready"}
{"level":"debug","ts":1632937783.824117,"logger":"ealinit","msg":"MainThread is running"}
{"level":"info","ts":1632937783.832964,"logger":"pktmbuf","msg":"decreasing Dataroom","key":"PAYLOAD","oldDataroom":11285,"newDataroom":9128}
{"level":"debug","ts":1632937783.8330622,"logger":"eal","msg":"object ID allocated","type":"ethvdev.Memif","id":"K0000000000000001"}
{"level":"debug","ts":1632937783.833236,"logger":"DPDK","msg":"rte_pmd_memif_probe(): Initialize MEMIF: net_memifK0000000000000001."}
{"level":"warn","ts":1632937783.8332891,"logger":"DPDK","msg":"rte_pmd_memif_probe(): Failed to register mp action callback: Operation not supported"}
{"level":"info","ts":1632937783.8333497,"logger":"eal","msg":"vdev initialized","name":"net_memifK0000000000000001","args":"socket-abstract=no,mac=F2:6D:65:6D:69:66,id=0,role=client,bsize=9000,rsize=10,socket=/run/ndn/fileserver.sock","socket":"any"}
{"level":"info","ts":1632937783.8334177,"logger":"ethface","msg":"port opened","port":0}
{"level":"debug","ts":1632937783.8336306,"logger":"eal","msg":"object ID allocated","type":"pktmbuf.Pool","id":"K0000000000000002"}
{"level":"debug","ts":1632937783.861061,"logger":"pktmbuf","msg":"mempool created","template":"DIRECT","socket":1,"use-socket":"any","cfg":{"capacity":65535,"privSize":480,"dataroom":9128},"pool":"K0000000000000002"}
{"level":"debug","ts":1632937783.861116,"logger":"eal","msg":"object ID allocated","type":"pktmbuf.Pool","id":"K0000000000000003"}
{"level":"debug","ts":1632937783.8676288,"logger":"pktmbuf","msg":"mempool created","template":"INDIRECT","socket":1,"use-socket":"any","cfg":{"capacity":65535,"privSize":0,"dataroom":0},"pool":"K0000000000000003"}
{"level":"debug","ts":1632937783.8676648,"logger":"eal","msg":"object ID allocated","type":"pktmbuf.Pool","id":"K0000000000000004"}
{"level":"debug","ts":1632937783.8836818,"logger":"pktmbuf","msg":"mempool created","template":"HEADER","socket":1,"use-socket":"any","cfg":{"capacity":65535,"privSize":480,"dataroom":186},"pool":"K0000000000000004"}
{"level":"debug","ts":1632937783.8837047,"logger":"eal","msg":"object ID allocated","type":"ringbuffer.Ring","id":"K0000000000000005"}
{"level":"debug","ts":1632937783.8837306,"logger":"eal","msg":"object ID allocated","type":"iface.Reassembler","id":"K0000000000000006"}
{"level":"debug","ts":1632937783.8837607,"logger":"eal","msg":"object ID allocated","type":"iface.Reassembler","id":"K0000000000000007"}
{"level":"debug","ts":1632937783.8837886,"logger":"eal","msg":"object ID allocated","type":"iface.Reassembler","id":"K0000000000000008"}
{"level":"debug","ts":1632937783.883813,"logger":"eal","msg":"object ID allocated","type":"iface.Reassembler","id":"K0000000000000009"}
{"level":"debug","ts":1632937783.8838398,"logger":"eal","msg":"object ID allocated","type":"iface.Reassembler","id":"K000000000000000a"}
{"level":"debug","ts":1632937783.8838615,"logger":"eal","msg":"object ID allocated","type":"iface.Reassembler","id":"K000000000000000b"}
{"level":"debug","ts":1632937783.883887,"logger":"eal","msg":"object ID allocated","type":"iface.Reassembler","id":"K000000000000000c"}
{"level":"debug","ts":1632937783.8839092,"logger":"eal","msg":"object ID allocated","type":"iface.Reassembler","id":"K000000000000000d"}
{"level":"info","ts":1632937783.883964,"logger":"ethface","msg":"impl initialized","port":0,"impl":"RxMemif"}
{"level":"debug","ts":1632937783.884003,"logger":"pktmbuf","msg":"mempool found","template":"DIRECT","socket":"any","use-socket":"any","cfg":{"capacity":65535,"privSize":480,"dataroom":9128},"pool":"K0000000000000002"}
{"level":"info","ts":1632937783.8841605,"logger":"ethdev","msg":"ethdev started","id":0,"name":"net_memifK0000000000000001","mtu":"unchanged","rxq":1,"txq":1,"promisc":false}
{"level":"info","ts":1632937783.8842008,"logger":"ethface","msg":"face started","port":0,"impl":"RxMemif","face":14996}
{"level":"info","ts":1632937783.8842337,"logger":"iface","msg":"face created","id":14996,"socket":1,"mtu":9014,"locator":{"rxQueueSize":64,"txQueueSize":64,"role":"client","socketName":"/run/ndn/fileserver.sock","id":0,"dataroom":9000,"ringCapacity":1024}}
{"level":"debug","ts":1632937783.8843315,"logger":"eal","msg":"object ID allocated","type":"ringbuffer.Ring","id":"K000000000000000e"}
{"level":"debug","ts":1632937783.8844244,"logger":"eal","msg":"object ID allocated","type":"pktmbuf.Pool","id":"K000000000000000f"}
{"level":"debug","ts":1632937783.8847501,"logger":"DPDK","msg":"memif_connect(): Connected."}
{"level":"debug","ts":1632937783.884798,"logger":"DPDK","msg":"memif_msg_receive_connected(): Remote interface net_memifK0000000000000049 connected."}
{"level":"debug","ts":1632937784.1268952,"logger":"pktmbuf","msg":"mempool created","template":"PAYLOAD","socket":1,"use-socket":"any","cfg":{"capacity":65535,"privSize":480,"dataroom":9128},"pool":"K000000000000000f"}
{"level":"debug","ts":1632937784.1269932,"logger":"eal","msg":"object ID allocated","type":"ringbuffer.Ring","id":"K0000000000000010"}
{"level":"debug","ts":1632937784.127061,"logger":"pktmbuf","msg":"mempool found","template":"PAYLOAD","socket":1,"use-socket":"any","cfg":{"capacity":65535,"privSize":480,"dataroom":9128},"pool":"K000000000000000f"}
```
Here is the FIB:
```bash
{"id":"4UAVPS3EI4V2IAAOPF4QI5EJ4SUCIRS9C8OG","name":"/8=ndnc/8=ft","nexthops":[{"id":"52DFVFAEQ463KCO7RR3AAJ46"}],"strategy":{"id":"DE7ULB2AHCNJOA81HEAV2LF47KMOE8UHCCUEQ6G"}}
```

and the fileserver configuration
```ts
import type { ActivateFileServerArgs } from "@usnistgov/ndn-dpdk";
import stdout from "stdout-stream";

const args: ActivateFileServerArgs = {
  eal: {
    coresPerNuma: { 0: 0, 1: 4 },
    memPerNuma: { 0: 0, 1: 4096 },
    filePrefix: "fileserver",
  },
  mempool: {
    DIRECT: { capacity: 65535, dataroom: 9128 },
    INDIRECT: { capacity: 65535 },
    PAYLOAD: { capacity: 65535, dataroom: 9128 },
  },
  face: {
    scheme: "memif",
    socketName: "/run/ndn/fileserver.sock",
    id: 0,
    role: "client",
    dataroom: 9000,
  },
  fileServer: {
    nThreads: 2,
    mounts: [
      { prefix: "/ndnc/ft", path: "/root/samples" }
    ],
    segmentLen: 6 * 1024,
    uringCapacity: 4096,
  },
};

stdout.write(JSON.stringify(args));
```

And this is how the Interest expressed by my consumer looks like
```bash
Request timeout for /ndnc/ft/root/samples/empty.dat/32=metadata
```

**Comment by Pesa** [comment]
With this configuration:

```js
mounts: [
      { prefix: "/ndnc/ft", path: "/root/samples" }
]
```

if the local file path is `/root/samples/empty.dat`, the metadata Interest name should be `/ndnc/ft/empty.dat/32=metadata`.

**Comment by yoursunny** [comment]
> I don't know if the Interest reaches the server.

You can look at face counters.
Run `ndndpdk-ctrl list-face` to discover face IDs, then `ndndpdk-ctrl get-face --cnt` to retrieve counters of a face.
This is available in both forwarder and file server.

You can see additional counters via GraphQL.
Use [Altair GraphQL Client](https://altair.sirmuel.design/) or another tool to see available counters in the schema.

> Are there any logs for this that I can activate?

You can see file server logs by setting environment variable `NDNDPDK_LOG_FileServer=V` and `NDNDPDK_LOG_FileServerFd=V`.
To change log level of a running file server, use GraphQL `setLogLevel` mutation.

**Comment by cataliniordache** [comment]
Thanks @Pesa. That however solves part of the issues, I still should have gotten a NACK packet from the server.

here are the face counters:
```bash
{"counters":{"rxData":"0","rxInterests":"0","rxNacks":"0","txData":"0","txInterests":"8","txNacks":"0"},"id":"52DFVFAEQ463KCO7RR3AAJ46","locator":{"dataroom":9000,"id":0,"ringCapacity":1024,"role":"server","rxQueueSize":64,"scheme":"memif","socketName":"/run/ndn/fileserver.sock","txQueueSize":64}}
```

**Comment by cataliniordache** [comment]
I will try also activating the fileserver logs as well

**Comment by yoursunny** [comment]
> I still should have gotten a NACK packet from the server.

Yes, the file server should reply with ContentType=Nack packet, for metadata requests only.

> here are the face counters:

Assuming you have a forwarder in additional to the file server, there are three faces:

* Forwarder's face toward consumer.
* Forwarder's face toward file server.
* File server's face toward forwarder.

Look at the counters in all three faces, and you'll have some idea on where the packets are getting lost.

**Comment by Pesa** [comment]
> Thanks @Pesa. That however solves part of the issues, I still should have gotten a NACK packet from the server.

Yes, good point.

**Comment by Pesa** [comment]
Given that the original issue "Unable to activate NDN-DPDK as fileserver" has been resolved, should we open a separate issue for the missing packets?

**Comment by cataliniordache** [comment]
@Pesa yes, I can close this issue. Tomorrow I will continue testing and will get counters from all faces on the forwarder side to further understand what is the problem. I will open another issue if needed. @yoursunny, @Pesa  thanks a lot for your help so far!
